From 42580fa54f890c6c21a7b2142b9c4c840f30eeb5 Mon Sep 17 00:00:00 2001
From: "Guoniu.Zhou" <guoniu.zhou@nxp.com>
Date: Wed, 6 Sep 2017 14:23:13 +0800
Subject: [PATCH 2518/5242] MLK-16374-3: PxP: improve stride parameter setting
 compatible

commit  67ba643b77260199e6eac24ad5e5d2d3a82502c3 from
https://source.codeaurora.org/external/imx/linux-imx.git

In pxp lib, the unit of stride parameter is pixel and stride
is not equal with width parameter of out buffer in some cases.

In order to use latest pxp lib in old version rootfs, PXP_DEVICE_LEGACY
macro is used to distinguish pxp drvier version. Because the
new pxp driver define a new variable and pxp lib can know this
through PXP_DEVICE_LEGACY, and determine if use it.

Signed-off-by: Guoniu.Zhou <guoniu.zhou@nxp.com>
Reviewed-by: Fancy Fang <chen.fang@nxp.com>
(cherry picked from commit 91da74e81cfdd2774b39a7574edf93de3f2a3f25)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/pxp/pxp_dma_v3.c |   16 ++++++++++++----
 include/uapi/linux/pxp_dma.h |    3 +++
 2 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/drivers/dma/pxp/pxp_dma_v3.c b/drivers/dma/pxp/pxp_dma_v3.c
index 633bcd9..c7c4332 100644
--- a/drivers/dma/pxp/pxp_dma_v3.c
+++ b/drivers/dma/pxp/pxp_dma_v3.c
@@ -198,6 +198,7 @@
 static dma_addr_t paddr;
 static bool v3p_flag;
 static int alpha_blending_version;
+static bool pxp_legacy;
 
 struct pxp_dma {
 	struct dma_device dma;
@@ -3618,10 +3619,15 @@ static int convert_param_to_pixmap(struct pxp_pixmap *pixmap,
 	pixmap->paddr  = param->paddr;
 	pixmap->bpp    = get_bpp_from_fmt(pixmap->format);
 
-	if (!param->stride || (param->stride == param->width))
-		pixmap->pitch  = param->width * pixmap->bpp >> 3;
-	else
-		pixmap->pitch = param->stride;
+	if (pxp_legacy) {
+		pixmap->pitch = (param->stride) ? (param->stride * pixmap->bpp >> 3) :
+						(param->width * pixmap->bpp >> 3);
+	} else {
+		if (!param->stride || (param->stride == param->width))
+			pixmap->pitch  = param->width * pixmap->bpp >> 3;
+		else
+			pixmap->pitch = param->stride;
+	}
 
 	pixmap->crop.x = param->crop.left;
 	pixmap->crop.y = param->crop.top;
@@ -3675,6 +3681,8 @@ static void __pxpdma_dostart(struct pxp_channel *pxp_chan)
 	else
 		alpha_blending_version = PXP_ALPHA_BLENDING_NONE;
 
+	pxp_legacy = (proc_data->pxp_legacy) ? true : false;
+
 	/* Save PxP configuration */
 	list_for_each_entry(child, &desc->tx_list, list) {
 		if (i == 0) {	/* Output */
diff --git a/include/uapi/linux/pxp_dma.h b/include/uapi/linux/pxp_dma.h
index 4b0b4f0..76983ab 100644
--- a/include/uapi/linux/pxp_dma.h
+++ b/include/uapi/linux/pxp_dma.h
@@ -127,6 +127,8 @@
 #define ALPHA_MODE_LEGACY	0x2
 #define ALPHA_MODE_PORTER_DUFF	0x3
 
+#define PXP_DEVICE_LEGACY
+
 /* Order significant! */
 enum pxp_channel_status {
 	PXP_CHANNEL_FREE,
@@ -313,6 +315,7 @@ struct pxp_proc_data {
 	bool reagl_en;		/* enable reagl/-d */
 	bool reagl_d_en;	/* enable reagl or reagl-d */
 	bool detection_only;
+	bool pxp_legacy;
 	int lut;
 	bool lut_cleanup;
 	unsigned int lut_status_1;
-- 
1.7.9.5

